src/rulesets/Base/array.jl

```diff
@@ -62,54 +62,43 @@ end
 #####
 ##### `repeat`
 #####
-
 function rrule(::typeof(repeat), xs::AbstractArray; inner=ntuple(_->1, ndims(xs)), outer=ntuple(_->1, ndims(xs)))
 
+    project_Xs = ProjectTo(xs)
+    S = size(xs)
     function repeat_pullback(ȳ)
         dY = unthunk(ȳ)
-        Δ′ = zero(xs)
-        S = size(xs)
-
-        # Loop through each element of Δ, calculate source dimensions, accumulate into Δ′
-        for (dest_idx, val) in pairs(IndexCartesian(), dY)
-            # First, round dest_idx[dim] to nearest gridpoint defined by inner[dim], then
-            # wrap around based on original size S.
-            src_idx = [mod1(div(dest_idx[dim] - 1, inner[dim]) + 1, S[dim]) for dim in 1:length(S)]
-            Δ′[src_idx...] += val
+        x̄ = @thunk begin
```

Comment:
> we shouldn't @thunk here because there is only one output

I was considering cases where it's part of the expression, but irrelevant to the gradient calculation. Consider any set-up where repeat is used, but just to set-up constants: 

```
foo(x) = identity(x)
function ChainRules.rrule(::typeof(foo), x)
    function foo_pullback(Ȳ)
        println("pb executed!")
        return (NoTangent(), Ȳ)
    end
    return foo(x), foo_pullback
end
```

```
Zygote.gradient(x -> identity(x), rand())
Zygote.gradient(x -> foo(x), rand()) # pb executed
Zygote.gradient(x -> foo(5) + identity(x), rand()) #pb executed
Zygote.gradient(x -> 5, 1.0)
```

Though doing this, I'm confused why the pb is executing in the third example. It seems that the pb execution is wasteful, given the primal evaluates to a constant. Think I'm misunderstanding something here.

Reply:
I don't think Zygote is smart enough to eliminate this kind of "dead code" in the backward pass. I don't think it can tell whether the argument (in this case a literal 5) has any dependence on x, so it just executes all of the pullbacks.

For the third example, I would expect it to run into: https://github.com/FluxML/Zygote.jl/blob/7f2b16987dc4eab236aabd889444cabdcc6f9caf/src/compiler/chainrules.jl#L143
so indeed I also find it weird why the pullback is executing.